juju/juju
# fix: Strange behavior of commands (juju status) after logout or login error

- Before this commit, the macaroon clearing logic in logout was not properly handled for existingController. Moreover, the way the accountDetails were provided to ClientAPI cause weird behavior when trying to call commands as not logged.
- After this commit, the macaroon clearing logic is centralized around the dial function which creates macaroon as a side effects, and the accountDetails may be nil when trying to call juju command unlogged (which was expected in some part of code but not on some others).

I tried to do the minimal changes in order to limit side effects. However, the login command is a mess and should be refactored.





## Checklist




## QA steps

Run two terminals after `make install` the code

First terminal
```sh
juju bootstrap lxd src --config identity-url=https://api.jujucharms.com/identity --config allow-model-access=true
juju change-user-password # change admin password
juju logout # cookie should be cleaned from value
juju status # an error should state that you are not logged in "src"
juju login -u notanuser@external # should cause browser auth through sso, then an error for mismatching notanuser and your account. Cookies should be clean and have no value
juju status # an error should state that you are not logged in "src"
juju login -u <your canonical account>@external # sso auth, you should be logged
juju add-model stuff # permission error
juju logout # logout
juju status # an error should state that you are not logged in "src"
juju login -u admin
juju grant <your canonical account>@external superuser  
juju logout
juju login -u <your canonical account>@external # sso auth,
juju status # show status
```

Second terminal
```sh
cd  ~/.local/share/juju/cookies
while true; do 
clear
cat src.json | jq '[.[] | select(.Path != "/identity/") | {Name, Updated, Expires, Value}]'                                                                                                                      
sleep 1s
done
```
On this terminal you should observe that the new macaroon added should have a value whenever you are logged, and no value whenever you logout / have a log error

## Documentation changes

*None*

## Links



**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/2072473

**Jira card:** [JUJU-6367](https://warthogs.atlassian.net/browse/JUJU-6367)
